test/Test/Database/LSMTree/Normal/StateMachine.hs

```diff
@@ -59,56 +74,133 @@ import qualified Test.QuickCheck.StateModel.Lockstep.Defaults as Lockstep.Defaul
+   QC.expectFailure $ -- TODO: remove once we have a real implementation
+    Lockstep.Run.runActionsBracket
+      (Proxy @(ModelState Impl.Real.TableHandle))
+      acquire
+      release
+      (\r (_, session) -> runReaderT r (session, handler))
+  where
+    acquire :: IO (FilePath, WrapSession Impl.Real.TableHandle IO)
+    acquire = do
+        (tmpDir, someHasFS) <- createSystemTempDirectory "propLockstepIO_RealIO"
+        session <- Impl.Real.newSession someHasFS (mkFsPath [])
+        pure (tmpDir, WrapSession session)
+
+    release :: (FilePath, WrapSession Impl.Real.TableHandle IO) -> IO ()
+    release (tmpDir, WrapSession session) = do
+        Impl.Real.closeSession session
+        removeDirectoryRecursive tmpDir
+
+    handler :: Handler IO (Maybe Model.Err)
+    handler = Handler $ pure . handler'
+      where
+        handler' :: IOError -> Maybe Model.Err
+        handler' _err = Nothing
+
+propLockstepIO_RealImpl_MockFS :: TestTree
+propLockstepIO_RealImpl_MockFS = testProperty "propLockstepIO_RealImpl_MockFS" $
+   QC.expectFailure $ -- TODO: remove once we have a real implementation
+    Lockstep.Run.runActionsBracket
+      (Proxy @(ModelState Impl.Real.TableHandle))
+      acquire
+      release
+      (\r session -> runReaderT r (session, handler))
+  where
+    acquire :: IO (WrapSession Impl.Real.TableHandle IO)
+    acquire = do
+        someHasFS <- SomeHasFS <$> simHasFS' MockFS.empty
+        WrapSession <$> Impl.Real.newSession someHasFS (mkFsPath [])
+
+    release :: WrapSession Impl.Real.TableHandle IO -> IO ()
+    release (WrapSession session) = Impl.Real.closeSession session
+
+    handler :: Handler IO (Maybe Model.Err)
+    handler = Handler $ pure . handler'
+      where
+        handler' :: IOError -> Maybe Model.Err
+        handler' _err = Nothing
+
+createSystemTempDirectory ::  [Char] -> IO (FilePath, SomeHasFS IO)
+createSystemTempDirectory prefix = do
     systemTempDir <- getCanonicalTemporaryDirectory
     tempDir <- createTempDirectory systemTempDir prefix
     pure (tempDir, SomeHasFS $ ioHasFS (MountPoint tempDir))
 
+instance Arbitrary Impl.ModelIO.TableConfig where
+  arbitrary :: Gen Impl.ModelIO.TableConfig
+  arbitrary = pure Impl.ModelIO.TableConfig
+
+-- TODO: improve, more types of configs
+instance Arbitrary Impl.Real.TableConfig where
+  arbitrary :: Gen Impl.Real.TableConfig
+  arbitrary = pure $  Impl.Real.TableConfig {
+        Impl.Real.tcMaxBufferMemory = 2 * 1024 * 1024
+      , Impl.Real.tcMaxBloomFilterMemory = 2 * 1024 * 1024
+      , Impl.Real.tcBitPrecision = ()
+      }
+
 {-------------------------------------------------------------------------------
   Model state
 -------------------------------------------------------------------------------}
 
-data ModelState = ModelState Model.Model Stats
+type ModelState :: ((Type -> Type) -> Type -> Type -> Type -> Type) -> Type
+data ModelState h = ModelState Model.Model Stats
```

Comment:
INFO: parameterising the model state by the table handle implementation `h` ensures that we can use the same statemachine for different table handle implementations. It does mean we've had to add a lot of `h` parameters everywhere.